backend/optimizer.py

```diff
@@ -1,46 +1,127 @@
-                    consumed_t += load_profile_utils.power(self.producer.schedule[i][1].load_profile, t)
-            produced_t = load_profile_utils.power(self.producer.prediction, t)
-
-            produced += produced_t
-            consumed += consumed_t
-            if produced_t - consumed_t < 0:
-                penalty += 1
-
-        score = math.fabs(produced - consumed) + penalty * 0
-        return score
+            for i, load in enumerate(self.differentiated_loads):
+                if schedule[i] > 0:
+                    consumed_t += load[t]
+
+            consumed.append(consumed_t)
+            produced.append(self.differentiated_production[t])
+
+        penalty = 0
+        diff = 0
+        for p, c in zip(produced, consumed):
+            diff_t = abs(p - c)
+            diff += diff_t
+            if c > p:
+                penalty += diff_t
+
+        return diff, penalty
+
+    def binary_optimzier(self):
+        configs = map(list, itertools.product([0, 1], repeat=len(self.differentiated_loads)))
+        min_config = None
+        min_config_value = float('inf')
+        for c in configs:
+            diff, penalty = self.to_minimize(c)
+            if diff < min_config_value:
+                min_config_value = diff + penalty
+                min_config = c
+        return min_config
+
+    def recursive_binary_optimizer(self, configs, min_value, min_config):
+        if len(configs) == 0:
+            return min_config
+        current_config = configs[0]
+        diff, penalty = self.to_minimize(current_config)
+
+        if penalty > min_value:
+            config_indices = [i for i, x in enumerate(current_config) if x == 1.0]
+            new_configs = list(filter(lambda c: not all(map(lambda i: c[i], config_indices)), configs))
+            return self.recursive_binary_optimizer(new_configs, min_value, min_config)
+
+        elif penalty + diff < min_value:
+            return self.recursive_binary_optimizer(configs[1:], penalty + diff, current_config)
+
+        else:
+            return self.recursive_binary_optimizer(configs[1:], min_value, min_config)
+
+    def differentiate_and_interpolate(self, series: pd.Series, indices: List[int]):
+        interpolated = self.interpolate(series, indices)
+        derivative = self.differentiate(interpolated)
+        return derivative
+
+    def differentiate(self, series: pd.Series):
+        indices = list(series.index)
+        values = list(series.values)
+        derivative = []
+        for i in range(1, len(indices) + 1):
+            if i >= len(indices):
+                derivative.insert(0, values[0])
+            else:
+                p0 = values[i - 1]
+                t0 = indices[i - 1]
+                p1 = values[i]
+                t1 = indices[i]
+
+                d = (p1 - p0)
+                derivative.append(d)
+
+        return pd.Series(index=indices, data=derivative)
+
+    def interpolate(self, series: pd.Series, indices: List[int]):
```

Comment:
First, find all values included in `indices` but not already having a value in `series`. Then add the value `NaN` for these new value, and interpolate with `pandas`. Last, we need to ensure that the values _before_ the series start is equal to the first value of the series, and that the values _after_ the series end is equal to the last value of the series -- or else, the functions would yield wrong values when optimizing.